From e9298122b6caf1deb4dbeb20e524b9076311ed66 Mon Sep 17 00:00:00 2001
From: PyScotch <pyscotch@noreply.invalid>
Date: Tue, 13 Oct 2026 13:52:45 +0000
Subject: [PATCH 090/240] Bugfix: free column block tree of compressed ordering
 on error in `hgraphOrderCp()` and `hmeshOrderCp()` [report R. Perry]

When the ordering strategy applied to the compressed graph (resp.
mesh) failed, only the arrays of the compressed structure were freed:
the column block tree that the failed strategy had already attached
to the coarse ordering, as well as its mutex, were leaked.

Call `orderExit()` on the coarse ordering on the error path, as is
done on the success path once the tree has been unlinked.

Refs: GRA-25
---
 src/libscotch/hgraph_order_cp.c | 3 ++-
 src/libscotch/hmesh_order_cp.c  | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/libscotch/hgraph_order_cp.c b/src/libscotch/hgraph_order_cp.c
index d7e0d66..fd60fcc 100644
--- a/src/libscotch/hgraph_order_cp.c
+++ b/src/libscotch/hgraph_order_cp.c
@@ -52,7 +52,7 @@
 /**                # Version 6.0  : from : 04 aug 2014     **/
 /**                                 to   : 27 jan 2020     **/
 /**                # Version 7.0  : from : 05 may 2019     **/
-/**                                 to   : 07 oct 2026     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /**   NOTES      : # Pre-hashing proves itself extremely   **/
 /**                  efficient, since for graphs that      **/
@@ -447,6 +447,7 @@ loop_failed: ;
   }
   orderInit (&coarordedat, coargrafdat.s.baseval, coargrafdat.vnohnbr, coarperitab); /* Build ordering of compressed subgraph */
   if (hgraphOrderSt (&coargrafdat, &coarordedat, 0, &coarordedat.rootdat, paraptr->stratcpr) != 0) {
+    orderExit  (&coarordedat);                    /* Free partially built column block tree */
     memFree    (coarperitab);
     hgraphExit (&coargrafdat);
     memFree    (finecoartax + finegrafptr->s.baseval);
diff --git a/src/libscotch/hmesh_order_cp.c b/src/libscotch/hmesh_order_cp.c
index 03cff06..8d31d92 100644
--- a/src/libscotch/hmesh_order_cp.c
+++ b/src/libscotch/hmesh_order_cp.c
@@ -46,7 +46,7 @@
 /**                # Version 6.0  : from : 06 jun 2018     **/
 /**                                 to   : 09 feb 2020     **/
 /**                # Version 7.0  : from : 28 aug 2020     **/
-/**                                 to   : 07 oct 2026     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /************************************************************/
 
@@ -384,6 +384,7 @@ loop_failed: ;
 
   orderInit (&coarordedat, coarmeshdat.m.baseval, coarmeshdat.m.vnodnbr, coarperitab); /* Build ordering of compressed submesh */
   if (hmeshOrderSt (&coarmeshdat, &coarordedat, 0, &coarordedat.rootdat, paraptr->stratcpr) != 0) {
+    orderExit (&coarordedat);                     /* Free partially built column block tree */
     hmeshExit (&coarmeshdat);
     return (1);
   }
-- 
2.47.3

